### PR TITLE
Portfolio

### PR DESCRIPTION
**Pull request by mateuszbaranczyk**

### CODE REVIEW

File: app/visualization.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+class Grapher:
+    def __init__(self) -> None:
+        self.exchange_rate_requester = ExchangeRateRequester()
+
+    def _get_historical_rates_df(self, transactions) -> DataFrame:
+        assert transactions, "There are no transactions"
+
+        first_transaction = transactions[0].date
+        historical_rates_date = self.exchange_rate_requester.get_historical_bids(first_transaction)
+        df = pd.DataFrame(list(historical_rates_date.items()), columns=["date", "rate"])
```
Comment (jacekplocharczyk) [inline app/visualization.py]: `df` -> `historical_rates_df`

File: app/visualization.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+        return df
+
+    def _get_operations_df(self, transactions) -> DataFrame:
+        assert transactions, "There are no transactions"
+
+        data_frame = self._get_historical_rates_df(transactions)
```
Comment (jacekplocharczyk) [inline app/visualization.py]: `data_frame` -> `operations_df`

File: app/visualization.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+        transaction_rate = transactions[0].rate
+        portfolio_value = transactions.pop(0).value
+
+        for i, row in data_frame.iterrows():
```
Comment (jacekplocharczyk) [inline app/visualization.py]: 1. copy whole and name it `_get_operations_df_smaller_list` (where you iterate over `transactions` not `historical_rates`
2. copy whole and name it `_get_operations_df_dataframe` (where you don't use `for` loop)

File: app/visualization.py (hunk 2 above)
Comment (jacekplocharczyk) [reply app/visualization.py]: dodaj kolumne buy_quota, sell_quota in EUR or one column with +/- sign

File: app/visualization.py (hunk 1 above)
Comment (mateuszbaranczyk) [reply app/visualization.py]: done

File: app/visualization.py (hunk 3 above)
Comment (mateuszbaranczyk) [reply app/visualization.py]: both - done

File: app/visualization.py (hunk 2 above)
Comment (mateuszbaranczyk) [reply app/visualization.py]: done

File: app/requester.py [hunk 4]
```diff
@@ -12,47 +13,38 @@ def get_rate(
...
-        )
+        api_url = self.api_url.format(currency_code=currency_code, start_date=start_date, end_date=end_date)
 
         response = requests.get(api_url).text
         data = json.loads(response)
         return data
 
     @staticmethod
-    def extract_todays_bid(rest_response):
+    def extract_todays_bid(rest_response) -> float:
```
Comment (jacekplocharczyk) [inline app/requester.py]: type hint missing for reset_response

File: app/requester.py [hunk 5]
```diff
@@ -12,47 +13,38 @@ def get_rate(
...
         todays_rates = rest_response["rates"][0]
         todays_bid = todays_rates["bid"]
         return todays_bid
 
     @staticmethod
-    def extract_historical_bids(rest_response):
+    def extract_historical_bids(rest_response) -> T.Dict[str, float]:
```
Comment (jacekplocharczyk) [inline app/requester.py]: type hint missing for reset_response

File: app/requester.py [hunk 6]
```diff
@@ -12,47 +13,38 @@ def get_rate(
...
         historical_rates = rest_response["rates"]
 
-        historical_bids = []
+        historical_bids = {}
         for rate in historical_rates:
             # historical_date = historical_rates[rate]
```
Comment (jacekplocharczyk) [inline app/requester.py]: please remove comment

File: app/requester.py [hunk 7]
```diff
@@ -12,47 +13,38 @@ def get_rate(
...
-    def get_todays_rate(
-        self,
-    ):  # ->float:
+    def get_todays_rate(self) -> float:
         rest_response = self.get_rate()
         todays_bid = self.extract_todays_bid(rest_response)
         return todays_bid
 
-    def get_historical_bids(self, start_date, end_date): # ->dict:
+    def get_historical_bids(self, start_date: str, end_date=str(datetime.date.today())) -> T.Dict[str, float]:
```
Comment (jacekplocharczyk) [inline app/requester.py]: `end_date=str(datetime.date.today())` -> `end_date: str=str(datetime.date.today())`

File: app/visualization.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
+from app.requester import ExchangeRateRequester
```
Comment (jacekplocharczyk) [inline app/visualization.py]: Application-specific imports should be at the bottom 
https://stackoverflow.com/a/42701457

File: app/visualization.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    def _create_historical_rates_df(self, transactions) -> pd.DataFrame:
+        assert transactions, "There are no transactions"
+
+        first_transaction = transactions[0].date
+        historical_rates_date = self.exchange_rate_requester.get_historical_bids(first_transaction)
+        historical_rates_df = pd.DataFrame(list(historical_rates_date.items()), columns=["date", "rate"])
+
+        return historical_rates_df
+
+    def _create_operations_df(self, transactions: list) -> pd.DataFrame:
```
Comment (jacekplocharczyk) [inline app/visualization.py]: you can extract some methods from this function ;)
It's quite long.

File: app/requester.py (hunk 6 above)
Comment (mateuszbaranczyk) [reply app/requester.py]: done

File: app/requester.py (hunk 7 above)
Comment (mateuszbaranczyk) [reply app/requester.py]: done

File: app/requester.py (hunk 4 above)
Comment (mateuszbaranczyk) [reply app/requester.py]: done

File: app/requester.py (hunk 5 above)
Comment (mateuszbaranczyk) [reply app/requester.py]: done

File: app/visualization.py (hunk 9 above)
Comment (mateuszbaranczyk) [reply app/visualization.py]: done

File: app/visualization.py [hunk 10]
```diff
@@ -16,13 +17,8 @@ def _create_historical_rates_df(self, transactions) -> pd.DataFrame:
...
 
-    def _create_operations_df(self, transactions: list) -> pd.DataFrame:
-        assert transactions, "There are no transactions"
-
-        operations_df = self._create_historical_rates_df(transactions)
-        transactions_df = pd.DataFrame(transactions, columns=["date", "transaction[+/-]", "transaction_rate"])
-        operations_df = pd.merge(operations_df, transactions_df, on="date", how="outer")
-
+    @staticmethod
+    def extract_create_operations_df(operations_df: pd.DataFrame) -> pd.DataFrame:
```
Comment (jacekplocharczyk) [inline app/visualization.py]: what means this name :P

Comment (mateuszbaranczyk) [reply app/visualization.py]: where aprove

**Review by jacekplocharczyk — APPROVED** [review]

**Pull request merged**